tendermint-proto: grpc server definitions for rpc/abci

erwanor opened the pull request:
<!--

Thanks for filing a PR!

Before hitting the button, please check the following items.  Please note that
every non-trivial PR must reference an issue that explains the changes in the
PR.

Please also make sure you've targeted the correct branch with your PR. See the
contributing guidelines for details.

-->

Partially addresses https://github.com/informalsystems/tendermint-rs/issues/1225. This PR adds gRPC server definitions for the rpc and abci services, but not (yet) for the privval service since it is blocked on #9256 for now. I have added @tony-iqlusion and @tomtau as coauthors since this PR is duplicating work that was previously included-then-yanked from `0.35`.

* [x] Referenced an issue explaining the need for the change
* [ ] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] Wrote tests
* [x] Added entry in `.changelog/`

mzabaluev commented on proto/Cargo.toml:
```diff
@@ -27,6 +33,7 @@ num-traits = { version = "0.2", default-features = false }
 num-derive = { version = "0.3", default-features = false }
 time = { version = "0.3", default-features = false, features = ["macros", "parsing"] }
 flex-error = { version = "0.4.4", default-features = false }
+tonic = { version = "0.8", optional = true }
```
Please update to tonic 0.9. Hermes already uses that version.

mzabaluev commented on proto/src/prost/v0_34/tendermint.rpc.grpc.rs:
```diff
@@ -18,3 +18,194 @@ pub struct ResponseBroadcastTx {
     #[prost(message, optional, tag = "2")]
     pub deliver_tx: ::core::option::Option<super::super::abci::ResponseDeliverTx>,
 }
+/// Generated server implementations.
```
If possible, avoid generating bindings for this package. It was removed from CometBFT as per https://github.com/cometbft/cometbft/issues/650

mzabaluev commented on proto/Cargo.toml:
```diff
@@ -13,6 +13,12 @@ description = """
+default = []
+grpc = ["grpc-server", "grpc-client"]
+grpc-server = ["tonic"]
+grpc-client = ["tonic"]
```
As this feature is not used by generated code, perhaps let's not introduce it at this point.

mzabaluev commented on proto/Cargo.toml:
```diff
@@ -13,6 +13,12 @@ description = """
 
+[features]
+default = []
+grpc = ["grpc-server", "grpc-client"]
```
Future-proofing for the eventual client bindings is good.

erwanor commented:
Thanks for the review @mzabaluev, I have bumped to `tonic@0.9`, removed the `grpc-client` feature, and modified the proto compiler to not generate bindings for the `rpc` server definitions.

However, I was not able to run the proto compiler as I previously did, so the PR has the old definitions:

```
⋊> ~/p/t/t/proto-compiler on tendermint-proto/grpc-server-traits ⨯ cargo run                                                                                                                                                                                                                                                                                                                                18:18:00
   Compiling tendermint-proto-compiler v0.3.0 (/Users/user/prod/tendermint-rs/tools/proto-compiler)
    Finished dev [unoptimized + debuginfo] target(s) in 0.88s
     Running `/Users/user/prod/tendermint-rs/tools/target/debug/tendermint-proto-compiler`
[info] => Fetching https://github.com/cometbft/cometbft at v0.34.28 into "/Users/user/prod/tendermint-rs/tools/proto-compiler/../target/tendermint"
  [info] => Cloning https://github.com/cometbft/cometbft into /Users/user/prod/tendermint-rs/tools/proto-compiler/../target/tendermint folder
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error { code: -1, klass: 2, message: "could not remove directory '/Users/user/prod/tendermint-rs/tools/target/tendermint/.git/refs/remotes/origin/pierre/fast-prototyping-1059': parent is not directory: Not a directory" }', proto-compiler/src/functions.rs:42:29
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

mzabaluev commented:
@erwanor This looks like a weird problem with the git checkout. Try removing `target/tendermint` and running the compiler again.

mzabaluev commented:
> modified the proto compiler to not generate bindings for the `rpc` server definitions.

They are still generated, but I have slapped the `grpc-server` attribute on them. Let's leave it at that, generating everything is simpler.

erwanor commented:
Thanks for your help with this, I owe you one!
